File: BinaryTree/java/levelAverages.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

class LevelAverages {
  public static List<Double> levelAverages(TreeNode root) {
    List<Double> result = new ArrayList<>();
    if (root == null) {
      return result;
    }

    Queue<TreeNode> queue = new ArrayDeque<>();
    queue.add(root);

    while (!queue.isEmpty()) {
      int levelSize = queue.size();
      double sum = 0;
      for (int i = 0; i < levelSize; i += 1) {
        TreeNode current = queue.remove();
        sum += current.val;
        if (current.left != null) queue.add(current.left);
        if (current.right != null) queue.add(current.right);
      }
      result.add(sum / levelSize);
    }

    return result;
  }

  public static void run() {
    // intentionally empty
  }
}

File: BinaryTree/python/levelAverages.py
from collections import deque
from typing import List, Optional

from node import Node


def level_averages(root: Optional<Node]) -> List[float]:
    if root is None:
        return []

    averages: List[float] = []
    queue = deque([root])
    while queue:
        level_sum = 0
        level_count = len(queue)
        for _ in range(level_count):
            current = queue.popleft()
            level_sum += current.val
            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
        averages.append(level_sum / level_count)
    return averages
